File: ui-tests-protractor/e2e/integrations/datamapper.steps.ts
import { binding, then, when } from 'cucumber-tsflow';
import { expect, P, World } from '../common/world';
import { DataMapperComponent } from './edit/datamapper.po';
import { ActionConfigureComponent, IntegrationEditPage } from '../integrations/edit/edit.po';
import { log } from '../../src/app/logging';

import { browser, ElementFinder, ExpectedConditions } from 'protractor';

@binding([World])
class DataMapperSteps {

  constructor(private world: World) {
  }

  @when(/^she creates mapping from "([^"]*)" to "([^"]*)"$/)
  public createMapping(source: string, target: string): P<any> {

    const mapper = new DataMapperComponent();
    return mapper.createMapping(source, target);
  }

  @then(/^she is presented with data mapper ui$/)
  public dataMapperUIpresent(): P<any> {

    const component = new DataMapperComponent();

    return expect(component.fieldsCount(), `data mapper ui must load and show fields count`)
      .to.eventually.be.greaterThan(5);
  }

  @when(/^she selects "([^"]*)" from "([^"]*)" selector-dropdown$/)
  public async selectFromDropDownByElement(option: string, selectAlias: string): P<any> {
    const datamapperCommon = new DataMapperComponent();
    const selectElement = await datamapperCommon.getSelectByAlias(selectAlias);
    await browser.wait(ExpectedConditions.visibilityOf(selectElement), 6000, `Input ${selectAlias} not loaded in time`);
    return this.world.app.selectOption(selectElement, option);
  }

  @then(/^she fills "([^"]*)" selector-input with "([^"]*)" value$/)
  public async fillActionConfigureField(selectorAlias: string, value: string): P<any> {

    const actionConf = new ActionConfigureComponent();
    // const datamapperComponent = new DataMapperComponent();
    const datamapperCommon = new DataMapperComponent();
    const inputElement = await datamapperCommon.getInputByAlias(selectorAlias);
    await browser.wait(ExpectedConditions.visibilityOf(inputElement), 6000, 'Input is not visible');
    return actionConf.fillInputByElement(inputElement, value);
  }

  /**
   * @param first parameter to be combined.
   * @param first_pos position of the first parameter in the final string
   * @param second parameter to be combined.
   * @param sec_pos position of the second parameter in the final string.
   * @param combined above two into this parameter.
   * @param separator used to estethically join first and second parameter.
   */
  // And she combines "FirstName" as "2" with "LastName" as "1" to "first_and_last_name" using "Space" separator
  @then(/^she combines "([^"]*)" as "([^"]*)" with "([^"]*)" as "([^"]*)" to "([^"]*)" using "([^"]*)" separator$/)
  public async combinePresentFielsWithAnother(first: string, first_pos: string, 
    second: string, sec_pos: string, combined: string, separator: string): P<any> {

    let inputElement: ElementFinder;
    let selectElement: ElementFinder;
    const world: World = new World();

    // Then she fills "FirstCombine" selector-input with "FirstName" value
    const actionConf = new ActionConfigureComponent();
    const datamapperCommon = new DataMapperComponent();
    inputElement = await datamapperCommon.getInputByAlias('FirstCombine');
    await browser.wait(ExpectedConditions.visibilityOf(inputElement), 6000, 'Input FirstCombineis not visible');
    await actionConf.fillInputByElement(inputElement, first);

    // And she selects "Combine" from "ActionSelect" selector-dropdown
    selectElement = await datamapperCommon.getSelectByAlias('ActionSelect');
    await browser.wait(ExpectedConditions.visibilityOf(selectElement), 6000, `Select ActionSelect not loaded in time`);
    await this.world.app.selectOption(selectElement, 'Combine');

    // And she selects "Space" from "SeparatorSelect" selector-dropdown
    selectElement = await datamapperCommon.getSelectByAlias('SeparatorSelect');
    await browser.wait(ExpectedConditions.visibilityOf(selectElement), 6000, `Select SeparatorSelect not loaded in time`);
    await this.world.app.selectOption(selectElement, separator);

    // And clicks on the "Add Source" link 
    await world.app.clickLink('Add Source');

    // Then she fills "SecondCombine" selector-input with "LastName" value
    inputElement = await datamapperCommon.getInputByAlias('SecondCombine');
    await browser.wait(ExpectedConditions.visibilityOf(inputElement), 6000, 'Input SecondCombine is not visible');
    await actionConf.fillInputByElement(inputElement, second);

    // And she fills "FirstCombinePosition" selector-input with "2" value
    inputElement = await datamapperCommon.getInputByAlias('FirstCombinePosition');
    await browser.wait(ExpectedConditions.visibilityOf(inputElement), 6000, 'Input FirstCombinePosition is not visible');
    await actionConf.fillInputByElement(inputElement, first_pos);

    // And she fills "SecondCombinePosition" selector-input with "1" value
    inputElement = await datamapperCommon.getInputByAlias('SecondCombinePosition');
    await browser.wait(ExpectedConditions.visibilityOf(inputElement), 6000, 'Input SecondCombinePosition is not visible');
    await actionConf.fillInputByElement(inputElement, sec_pos);

    // Then she fills "TargetCombine" selector-input with "first_and_last_name" value
    inputElement = await datamapperCommon.getInputByAlias('TargetCombine');
    await browser.wait(ExpectedConditions.visibilityOf(inputElement), 6000, 'Input TargetCombine is not visible');
    return actionConf.fillInputByElement(inputElement, combined);

  }
}

export = DataMapperSteps;

File: ui-tests/src/test/java/io/syndesis/qe/steps/integrations/DataMapperSteps.java
package io.syndesis.qe.steps.integrations;

import static org.junit.Assert.assertThat;

import static org.hamcrest.Matchers.greaterThan;

import static com.codeborne.selenide.Condition.visible;

import com.codeborne.selenide.SelenideElement;

import cucumber.api.java.en.Then;
import cucumber.api.java.en.When;
import io.syndesis.qe.pages.integrations.edit.IntegrationEditPage;
import io.syndesis.qe.pages.integrations.edit.steps.DataMapperComponent;
import lombok.extern.slf4j.Slf4j;

/**
 * Created by sveres on 11/15/17.
 */
@Slf4j
public class DataMapperSteps {

	private DataMapperComponent mapper = new IntegrationEditPage().getDataMapperComponent();

	@When("^she creates mapping from \"([^\"]*)\" to \"([^\"]*)\"$")
	public void createMapping(String source, String target) {
		mapper.createMapping(source, target);
	}

	@Then("^she is presented with data mapper ui$")
	public void dataMapperUIpresent() {
		log.info("data mapper ui must load and show fields count");
		assertThat(mapper.fieldsCount(), greaterThan(5));
	}

	@When("^she selects \"([^\"]*)\" from \"([^\"]*)\" selector-dropdown$")
	public void selectFromDropDownByElement(String option, String selectAlias) {
		log.info(option);
		SelenideElement selectElement = mapper.getElementByAlias(selectAlias).shouldBe(visible);
		mapper.selectOption(selectElement, option);
	}

	@Then("^she fills \"([^\"]*)\" selector-input with \"([^\"]*)\" value$")
	public void fillActionConfigureField(String selectorAlias, String value) {
		SelenideElement inputElement = mapper.getElementByAlias(selectorAlias).shouldBe(visible);
		mapper.fillInput(inputElement, value);
	}

	/**
	 * @param first parameter to be combined.
	 * @param first_pos position of the first parameter in the final string
	 * @param second parameter to be combined.
	 * @param sec_pos position of the second parameter in the final string.
	 * @param combined above two into this parameter.
	 * @param separator used to estethically join first and second parameter.
	 */
	// And she combines "FirstName" as "2" with "LastName" as "1" to "first_and_last_name" using "Space" separator
	@Then("^she combines \"(\\w+)\" as \"(\\w+)\" with \"(\\w+)\" as \"(\\w+)\" to \"(\\w+)\" using \"(\\w+)\" separator$")
	public void combinePresentFielsWithAnother(String first, String first_pos,
			String second, String sec_pos, String combined, String separator) {
		SelenideElement inputElement;
		SelenideElement selectElement;

		// Then she fills "FirstCombine" selector-input with "FirstName" value
		inputElement = mapper.getElementByAlias("FirstSource").shouldBe(visible);
		mapper.fillInput(inputElement, first);

		// And she selects "Combine" from "ActionSelect" selector-dropdown
		selectElement = mapper.getElementByAlias("ActionSelect").shouldBe(visible);
		mapper.selectOption(selectElement, "Combine");

		// And she selects "Space" from "SeparatorSelect" selector-dropdown
		selectElement = mapper.getElementByAlias("SeparatorSelect").shouldBe(visible);
		mapper.selectOption(selectElement, separator);

		// And clicks on the "Add Source" link
		mapper.clickLink("Add Source");

		// Then she fills "SecondCombine" selector-input with "LastName" value
		inputElement = mapper.getElementByAlias("SecondSource").shouldBe(visible);
		mapper.fillInput(inputElement, second);

		// And she fills "FirstCombinePosition" selector-input with "2" value
		inputElement = mapper.getElementByAlias("FirstSourcePosition").shouldBe(visible);
		mapper.fillInput(inputElement, first_pos);

		// And she fills "SecondCombinePosition" selector-input with "1" value
		inputElement = mapper.getElementByAlias("SecondSourcePosition").shouldBe(visible);
		mapper.fillInput(inputElement, sec_pos);

		// Then she fills "TargetCombine" selector-input with "first_and_last_name" value
		inputElement = mapper.getElementByAlias("FirstTarget").shouldBe(visible);
		mapper.fillInput(inputElement, combined);
	}

	//	And she separates "FirstName" into "company" as "2" and "email" as "1" using "Comma" separator
	@Then("^she separates \"(\\w+)\" into \"(\\w+)\" as \"(\\w+)\" and \"(\\w+)\" as \"(\\w+)\" using \"(\\w+)\" separator$")
	public void separatePresentFielsIntoTwo(String input, String output1, String first_pos, String output2, String second_pos, String separator) {
		SelenideElement inputElement;
		SelenideElement selectElement;

		inputElement = mapper.getElementByAlias("FirstSource").shouldBe(visible);
		mapper.fillInput(inputElement, input);

		selectElement = mapper.getElementByAlias("ActionSelect").shouldBe(visible);
		mapper.selectOption(selectElement, "Separate");

		selectElement = mapper.getElementByAlias("SeparatorSelect").shouldBe(visible);
		mapper.selectOption(selectElement, separator);

		// NOTE: THIS STEP SHOULD HAVE BEEN DONE AUTOMATICALLY BY SELECTING "Separate" action
		mapper.clickLink("Add Target");

		inputElement = mapper.getElementByAlias("FirstTarget").shouldBe(visible);
		mapper.fillInput(inputElement, output1);

		inputElement = mapper.getElementByAlias("FirstTargetPosition").shouldBe(visible);
		mapper.fillInput(inputElement, first_pos);

		inputElement = mapper.getElementByAlias("SecondTarget").shouldBe(visible);
		mapper.fillInput(inputElement, output2);

		inputElement = mapper.getElementByAlias("SecondTargetPosition").shouldBe(visible);
		mapper.fillInput(inputElement, second_pos);

	}
}
